File: cld/src/main/java/cloudcmd/cld/commands/Sync.java
package cloudcmd.cld.commands;


import cloudcmd.cld.CloudEngineService;
import cloudcmd.common.JsonUtil;
import cloudcmd.cld.IndexStorageService;
import jpbetz.cli.Command;
import jpbetz.cli.CommandContext;
import jpbetz.cli.Opt;
import jpbetz.cli.SubCommand;
import org.json.JSONArray;
import org.json.JSONObject;

@SubCommand(name = "push", description = "Sync the local cache to storage endpoints.")
public class Sync implements Command
{
  @Opt(opt = "n", longOpt = "minTier", description = "min tier to verify to", required = false)
  Number _minTier = 0;

  @Opt(opt = "m", longOpt = "maxTier", description = "max tier to verify to", required = false)
  Number _maxTier = Integer.MAX_VALUE;

  @Opt(opt = "a", longOpt = "all", description = "sync all", required = false)
  boolean _syncAll = true;

  @Override
  public void exec(CommandContext commandLine) throws Exception
  {
    JSONArray selections = _syncAll ? IndexStorageService.instance().find(new JSONObject()) : JsonUtil.loadJsonArray(System.in);
    System.err.println(String.format("syncing %d files", selections.length()));
    CloudEngineService.instance().sync(_minTier.intValue(), _maxTier.intValue(), selections);
  }
}

File: cld/src/main/scala/cloudcmd/cld/commands/Sync.scala
package cloudcmd.cld.commands

import cloudcmd.common.util.{FileMetaData, JsonUtil}
import jpbetz.cli.Command
import jpbetz.cli.CommandContext
import jpbetz.cli.Opt
import jpbetz.cli.SubCommand
import cloudcmd.cld.CloudServices
import org.json.JSONArray

@SubCommand(name = "sync", description = "Sync the local cache to storage endpoints.")
class Sync extends Command {

  @Opt(opt = "n", longOpt = "minTier", description = "min tier to verify to", required = false) private var _minTier: Number = 0
  @Opt(opt = "m", longOpt = "maxTier", description = "max tier to verify to", required = false) private var _maxTier: Number = Integer.MAX_VALUE
  @Opt(opt = "a", longOpt = "all", description = "sync all", required = false) private var _syncAll: Boolean = true

  def exec(commandLine: CommandContext) {

    val selections = if (_syncAll) {
      val arr = new JSONArray
      val metaHashes = CloudServices.CloudEngine.getMetaHashSet()
      val fmds = metaHashes.par.map{ hash => FileMetaData.create(hash, JsonUtil.loadJson(CloudServices.CloudEngine.load(hash))) }.toList
      fmds.foreach(a => arr.put(a.toJson))
      arr
    } else {
      JsonUtil.loadJsonArray(System.in)
    }

    System.err.println("syncing %d files".format(selections.length))
    CloudServices.CloudEngine.filterAdapters(_minTier.intValue, _maxTier.intValue)
    CloudServices.IndexStorage.sync(selections)
    System.err.println("rebuilding file search index...")
    CloudServices.IndexStorage.reindex
  }
}